File: src-tauri/src/repo/notifications.rs
//! The outbox.
//!
//! A row is written before anything is sent. `UNIQUE (rule_id, policy_id,
//! policy_period)` is what makes a reminder fire exactly once per policy year,
//! however many times the sweep runs or the app restarts, and it is why a
//! crash mid-send costs at most one duplicate rather than a mailshot.

use rusqlite::{params, types::Value, Connection};

use crate::error::{AppError, AppResult};
use crate::models::{Notification, NotificationFilter, Page, NOTIFICATION_COLUMNS};
use crate::query::{self, Conditions};

pub const STATUSES: &[&str] = &["queued", "sent", "failed", "skipped", "cancelled"];

const SORTABLE: &[(&str, &str)] = &[
    ("scheduledFor", "n.scheduled_for"),
    ("createdAt", "n.created_at"),
    ("sentAt", "n.sent_at"),
    ("status", "n.status"),
    ("clientName", "client_name"),
];

/// What goes into the outbox before sending.
pub struct NewNotification {
    pub rule_id: i64,
    pub policy_id: i64,
    pub client_id: i64,
    pub policy_period: String,
    pub audience: String,
    pub channel: String,
    pub to_address: Option<String>,
    pub subject: String,
    pub body: String,
    pub scheduled_for: String,
}

/// The message itself, fetched only when a row is about to be sent — the
/// snapshots are large and no list needs them.
pub struct Payload {
    pub id: i64,
    pub channel: String,
    pub to_address: Option<String>,
    pub client_name: String,
    pub subject: String,
    pub body: String,
}

/// Returns the new row id, or None when this reminder already exists for this
/// policy year and nothing was written.
pub fn queue(conn: &Connection, entry: &NewNotification) -> AppResult<Option<i64>> {
    let changed = conn.execute(
        "INSERT OR IGNORE INTO notification_log (rule_id, policy_id, client_id, policy_period, \
             audience, channel, to_address, subject, body_snapshot, status, scheduled_for) \
         VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9, 'queued', ?10)",
        params![
            entry.rule_id,
            entry.policy_id,
            entry.client_id,
            entry.policy_period,
            entry.audience,
            entry.channel,
            entry.to_address,
            entry.subject,
            entry.body,
            entry.scheduled_for,
        ],
    )?;
    Ok(if changed == 0 {
        None
    } else {
        Some(conn.last_insert_rowid())
    })
}

/// Whether this rule has already been recorded against this policy year, in any
/// state. Used by the planner so a preview does not offer what will be ignored.
pub fn already_logged(
    conn: &Connection,
    rule_id: i64,
    policy_id: i64,
    period: &str,
) -> AppResult<bool> {
    let count: i64 = conn.query_row(
        "SELECT COUNT(*) FROM notification_log \
         WHERE rule_id = ?1 AND policy_id = ?2 AND policy_period = ?3",
        params![rule_id, policy_id, period],
        |row| row.get(0),
    )?;
    Ok(count > 0)
}

/// Queued rows whose time has come, oldest first so a backlog drains in order.
pub fn due(conn: &Connection, now_iso: &str, limit: i64) -> AppResult<Vec<Payload>> {
    let mut stmt = conn.prepare(
        "SELECT n.id, n.channel, n.to_address, \
                COALESCE((SELECT c.full_name FROM clients c WHERE c.id = n.client_id), ''), \
                COALESCE(n.subject, ''), COALESCE(n.body_snapshot, '') \
         FROM notification_log n \
         WHERE n.status = 'queued' AND n.scheduled_for <= ?1 \
         ORDER BY n.scheduled_for, n.id LIMIT ?2",
    )?;
    let rows = stmt
        .query_map(params![now_iso, limit], |row| {
            Ok(Payload {
                id: row.get(0)?,
                channel: row.get(1)?,
                to_address: row.get(2)?,
                client_name: row.get(3)?,
                subject: row.get(4)?,
                body: row.get(5)?,
            })
        })?
        .collect::<rusqlite::Result<Vec<_>>>()?;
    Ok(rows)
}

pub fn mark_sent(conn: &Connection, id: i64) -> AppResult<()> {
    conn.execute(
        "UPDATE notification_log SET status = 'sent', attempts = attempts + 1, \
             last_error = NULL, sent_at = datetime('now') WHERE id = ?1",
        params![id],
    )?;
    Ok(())
}

/// Stays `queued` for the first few attempts so the next sweep picks it up; a
/// server that is down for an hour should not need the operator to intervene.
pub fn mark_attempt_failed(
    conn: &Connection,
    id: i64,
    error: &str,
    max_attempts: i64,
) -> AppResult<()> {
    conn.execute(
        "UPDATE notification_log \
         SET attempts = attempts + 1, last_error = ?2, \
             status = CASE WHEN attempts + 1 >= ?3 THEN 'failed' ELSE 'queued' END \
         WHERE id = ?1",
        params![id, error, max_attempts],
    )?;
    Ok(())
}

pub fn mark_skipped(conn: &Connection, id: i64, reason: &str) -> AppResult<()> {
    conn.execute(
        "UPDATE notification_log SET status = 'skipped', last_error = ?2 WHERE id = ?1",
        params![id, reason],
    )?;
    Ok(())
}

/// Puts a failed or cancelled row back in the queue for the next sweep.
pub fn requeue(conn: &Connection, id: i64) -> AppResult<()> {
    let changed = conn.execute(
        "UPDATE notification_log SET status = 'queued', attempts = 0, last_error = NULL, \
             scheduled_for = datetime('now') WHERE id = ?1 AND status IN ('failed', 'cancelled', 'skipped')",
        params![id],
    )?;
    if changed == 0 {
        return Err(AppError::Conflict(
            "Only a failed, skipped or cancelled reminder can be sent again".into(),
        ));
    }
    Ok(())
}

pub fn cancel(conn: &Connection, id: i64) -> AppResult<()> {
    let changed = conn.execute(
        "UPDATE notification_log SET status = 'cancelled' WHERE id = ?1 AND status = 'queued'",
        params![id],
    )?;
    if changed == 0 {
        return Err(AppError::Conflict(
            "Only a queued reminder can be cancelled".into(),
        ));
    }
    Ok(())
}

pub fn list(conn: &Connection, filter: &NotificationFilter) -> AppResult<Page<Notification>> {
    let mut conditions = Conditions::new();

    if let Some(statuses) = &filter.statuses {
        if let Some((clause, values)) = query::in_clause("n.status", statuses, STATUSES) {
            conditions.add_many(clause, values);
        }
    }
    if let Some(client_id) = filter.client_id {
        conditions.add("n.client_id = ?", client_id);
    }
    if let Some(policy_id) = filter.policy_id {
        conditions.add("n.policy_id = ?", policy_id);
    }
    if let Some(search) = filter.search.as_ref().filter(|s| !s.trim().is_empty()) {
        let pattern = query::like_pattern(search);
        conditions.add_many(
            "(n.to_address LIKE ? ESCAPE '\\' OR n.subject LIKE ? ESCAPE '\\' \
              OR EXISTS (SELECT 1 FROM clients c WHERE c.id = n.client_id \
                         AND c.full_name LIKE ? ESCAPE '\\'))"
                .to_string(),
            vec![
                Value::Text(pattern.clone()),
                Value::Text(pattern.clone()),
                Value::Text(pattern),
            ],
        );
    }

    let where_sql = conditions.where_sql();
    let total = super::count(
        conn,
        &format!("SELECT COUNT(*) FROM notification_log n{where_sql}"),
        conditions.params(),
    )?;

    let (page, page_size, limit, offset) = query::paginate(filter.page, filter.page_size);
    let order = query::order_by(
        filter.sort.as_deref(),
        filter.descending.unwrap_or(true),
        SORTABLE,
        "n.scheduled_for",
    );
    let sql = format!(
        "SELECT {NOTIFICATION_COLUMNS} FROM notification_log n{where_sql}{order} LIMIT ? OFFSET ?"
    );

    let mut stmt = conn.prepare(&sql)?;
    let rows = stmt
        .query_map(
            rusqlite::params_from_iter(conditions.params_with([limit, offset])),
            Notification::from_row,
        )?
        .collect::<rusqlite::Result<Vec<_>>>()?;

    Ok(Page {
        rows,
        total,
        page,
        page_size,
    })
}

pub fn count_by_status(conn: &Connection, status: &str) -> AppResult<i64> {
    let count = conn.query_row(
        "SELECT COUNT(*) FROM notification_log WHERE status = ?1",
        params![status],
        |row| row.get(0),
    )?;
    Ok(count)
}

pub fn sent_on(conn: &Connection, day_iso: &str) -> AppResult<i64> {
    let count = conn.query_row(
        "SELECT COUNT(*) FROM notification_log WHERE status = 'sent' AND date(sent_at) = ?1",
        params![day_iso],
        |row| row.get(0),
    )?;
    Ok(count)
}

/// Clears rows for a policy year that has since been renewed, so a reminder
/// queued yesterday does not chase a client who has already renewed.
pub fn cancel_for_policy(conn: &Connection, policy_id: i64) -> AppResult<usize> {
    let changed = conn.execute(
        "UPDATE notification_log SET status = 'cancelled', \
             last_error = 'The policy was renewed' \
         WHERE policy_id = ?1 AND status = 'queued'",
        params![policy_id],
    )?;
    Ok(changed)
}

File: legacy-windows/src/core/repo/notifications.ts
/**
 * A port of `src-tauri/src/repo/notifications.rs`: the outbox.
 *
 * A row is written before anything is sent. `UNIQUE (rule_id, policy_id,
 * policy_period)` is what makes a reminder fire exactly once per policy year,
 * however many times the sweep runs or the app restarts, and it is why a crash
 * mid-send costs at most one duplicate rather than a mailshot.
 */

import type { Conn } from "../db";
import { AppError } from "../errors";
import { Conditions, inClause, likePattern, orderBy, paginate } from "../query";
import { toModel, toModels } from "../rows";
import type { Notification, NotificationFilter, Page } from "../types";
import { count } from "./shared";

export const STATUSES = ["queued", "sent", "failed", "skipped", "cancelled"] as const;

const SORTABLE: Record<string, string> = {
  scheduledFor: "n.scheduled_for",
  createdAt: "n.created_at",
  sentAt: "n.sent_at",
  status: "n.status",
  clientName: "client_name",
};

const COLUMNS =
  "n.id, n.rule_id, " +
  "(SELECT r.name FROM reminder_rules r WHERE r.id = n.rule_id) AS rule_name, " +
  "n.policy_id, (SELECT p.policy_number FROM policies p WHERE p.id = n.policy_id) AS policy_number, " +
  "n.client_id, (SELECT c.full_name FROM clients c WHERE c.id = n.client_id) AS client_name, " +
  "n.policy_period, n.audience, n.channel, n.to_address, n.subject, n.status, n.attempts, " +
  "n.last_error, n.scheduled_for, n.sent_at, n.created_at";

/** What goes into the outbox before sending. */
export interface NewNotification {
  ruleId: number;
  policyId: number;
  clientId: number;
  policyPeriod: string;
  audience: string;
  channel: string;
  toAddress: string | null;
  subject: string;
  body: string;
  scheduledFor: string;
}

/**
 * The message itself, fetched only when a row is about to be sent — the snapshots
 * are large and no list needs them.
 */
export interface Payload {
  id: number;
  channel: string;
  toAddress: string | null;
  clientName: string;
  subject: string;
  body: string;
}

/**
 * Returns the new row id, or null when this reminder already exists for this
 * policy year and nothing was written.
 */
export function queue(conn: Conn, entry: NewNotification): number | null {
  const result = conn
    .prepare(
      "INSERT OR IGNORE INTO notification_log (rule_id, policy_id, client_id, policy_period, " +
        "audience, channel, to_address, subject, body_snapshot, status, scheduled_for) " +
        "VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, 'queued', ?)",
    )
    .run(
      entry.ruleId,
      entry.policyId,
      entry.clientId,
      entry.policyPeriod,
      entry.audience,
      entry.channel,
      entry.toAddress,
      entry.subject,
      entry.body,
      entry.scheduledFor,
    );
  return result.changes === 0 ? null : Number(result.lastInsertRowid);
}

/**
 * Whether this rule has already been recorded against this policy year, in any
 * state. Used by the planner so a preview does not offer what will be ignored.
 */
export function alreadyLogged(conn: Conn, ruleId: number, policyId: number, period: string): boolean {
  const row = conn
    .prepare(
      "SELECT COUNT(*) AS n FROM notification_log " +
        "WHERE rule_id = ? AND policy_id = ? AND policy_period = ?",
    )
    .get(ruleId, policyId, period) as { n: number };
  return row.n > 0;
}

/** Queued rows whose time has come, oldest first so a backlog drains in order. */
export function due(conn: Conn, nowIso: string, limit: number): Payload[] {
  const rows = conn
    .prepare(
      "SELECT n.id, n.channel, n.to_address, " +
        "COALESCE((SELECT c.full_name FROM clients c WHERE c.id = n.client_id), '') AS client_name, " +
        "COALESCE(n.subject, '') AS subject, COALESCE(n.body_snapshot, '') AS body " +
        "FROM notification_log n " +
        "WHERE n.status = 'queued' AND n.scheduled_for <= ? " +
        "ORDER BY n.scheduled_for, n.id LIMIT ?",
    )
    .all(nowIso, limit) as Record<string, unknown>[];
  return rows.map((row) => toModel<Payload>(row));
}

export function markSent(conn: Conn, id: number): void {
  conn
    .prepare(
      "UPDATE notification_log SET status = 'sent', attempts = attempts + 1, " +
        "last_error = NULL, sent_at = datetime('now') WHERE id = ?",
    )
    .run(id);
}

/**
 * Stays `queued` for the first few attempts so the next sweep picks it up; a
 * server that is down for an hour should not need the operator to intervene.
 */
export function markAttemptFailed(conn: Conn, id: number, error: string, maxAttempts: number): void {
  conn
    .prepare(
      "UPDATE notification_log " +
        "SET attempts = attempts + 1, last_error = ?, " +
        "    status = CASE WHEN attempts + 1 >= ? THEN 'failed' ELSE 'queued' END " +
        "WHERE id = ?",
    )
    .run(error, maxAttempts, id);
}

export function markSkipped(conn: Conn, id: number, reason: string): void {
  conn.prepare("UPDATE notification_log SET status = 'skipped', last_error = ? WHERE id = ?").run(reason, id);
}

/** Puts a failed or cancelled row back in the queue for the next sweep. */
export function requeue(conn: Conn, id: number): void {
  const result = conn
    .prepare(
      "UPDATE notification_log SET status = 'queued', attempts = 0, last_error = NULL, " +
        "scheduled_for = datetime('now') WHERE id = ? AND status IN ('failed', 'cancelled', 'skipped')",
    )
    .run(id);
  if (result.changes === 0) {
    throw AppError.conflict("Only a failed, skipped or cancelled reminder can be sent again");
  }
}

export function cancel(conn: Conn, id: number): void {
  const result = conn
    .prepare("UPDATE notification_log SET status = 'cancelled' WHERE id = ? AND status = 'queued'")
    .run(id);
  if (result.changes === 0) throw AppError.conflict("Only a queued reminder can be cancelled");
}

export function list(conn: Conn, filter: NotificationFilter): Page<Notification> {
  const conditions = new Conditions();

  if (filter.statuses) {
    const statuses = inClause("n.status", filter.statuses, STATUSES);
    if (statuses) conditions.addMany(statuses.clause, statuses.params);
  }
  if (filter.clientId != null) conditions.add("n.client_id = ?", filter.clientId);
  if (filter.policyId != null) conditions.add("n.policy_id = ?", filter.policyId);

  const search = filter.search?.trim();
  if (search) {
    const pattern = likePattern(search);
    conditions.addMany(
      "(n.to_address LIKE ? ESCAPE '\\' OR n.subject LIKE ? ESCAPE '\\'" +
        " OR EXISTS (SELECT 1 FROM clients c WHERE c.id = n.client_id" +
        "            AND c.full_name LIKE ? ESCAPE '\\'))",
      [pattern, pattern, pattern],
    );
  }

  const whereSql = conditions.whereSql();
  const total = count(
    conn,
    `SELECT COUNT(*) AS n FROM notification_log n${whereSql}`,
    conditions.params(),
  );

  const { page, pageSize, limit, offset } = paginate(filter.page, filter.pageSize);
  // Newest first unless asked otherwise: the outbox is read to find out what just
  // happened far more often than what happened first.
  const order = orderBy(filter.sort, filter.descending ?? true, SORTABLE, "n.scheduled_for");

  const rows = conn
    .prepare(`SELECT ${COLUMNS} FROM notification_log n${whereSql}${order} LIMIT ? OFFSET ?`)
    .all(...conditions.paramsWith([limit, offset])) as Record<string, unknown>[];

  return { rows: toModels<Notification>(rows), total, page, pageSize };
}

export function countByStatus(conn: Conn, status: string): number {
  const row = conn
    .prepare("SELECT COUNT(*) AS n FROM notification_log WHERE status = ?")
    .get(status) as { n: number };
  return row.n;
}

export function sentOn(conn: Conn, dayIso: string): number {
  const row = conn
    .prepare("SELECT COUNT(*) AS n FROM notification_log WHERE status = 'sent' AND date(sent_at) = ?")
    .get(dayIso) as { n: number };
  return row.n;
}

/**
 * Clears rows for a policy year that has since been renewed, so a reminder queued
 * yesterday does not chase a client who has already renewed.
 */
export function cancelForPolicy(conn: Conn, policyId: number): number {
  const result = conn
    .prepare(
      "UPDATE notification_log SET status = 'cancelled', " +
        "last_error = 'The policy was renewed' " +
        "WHERE policy_id = ? AND status = 'queued'",
    )
    .run(policyId);
  return result.changes;
}
